capa/features/extractors/binexport2/__init__.py

```diff
@@ -0,0 +1,309 @@
+    """
+
+    def filename_similarity_key(p: Path):
+        # note closure over input_file.
+        # sort first by length of common prefix, then by name (for stability)
+        return (compute_common_prefix_length(p.name, input_file.name), p.name)
+
+    wanted_sha256 = be2.meta_information.executable_id.lower()
+
+    input_directory = input_file.parent
+    siblings = [p for p in input_directory.iterdir() if p.is_file()]
+    siblings.sort(key=filename_similarity_key, reverse=True)
+    for sibling in siblings:
+        # e.g. with open IDA files in the same directory on Windows
+        with contextlib.suppress(PermissionError):
+            if hashlib.sha256(sibling.read_bytes()).hexdigest().lower() == wanted_sha256:
+                return sibling
+
+    for search_path in search_paths:
+        candidates = [p for p in search_path.iterdir() if p.is_file()]
+        candidates.sort(key=filename_similarity_key, reverse=True)
+        for candidate in candidates:
+            with contextlib.suppress(PermissionError):
+                if hashlib.sha256(candidate.read_bytes()).hexdigest().lower() == wanted_sha256:
+                    return candidate
+
+    raise ValueError("cannot find sample")
+
+
+class BinExport2Index:
+    def __init__(self, be2: BinExport2):
+        self.be2 = be2
+
+        self.callers_by_vertex_index: Dict[int, List[int]] = defaultdict(list)
+        self.callees_by_vertex_index: Dict[int, List[int]] = defaultdict(list)
+
+        # note: flow graph != call graph (vertex)
+        self.flow_graph_index_by_address: Dict[int, int] = {}
+        self.flow_graph_address_by_index: Dict[int, int] = {}
+
+        # edges that come from the given basic block
+        self.source_edges_by_basic_block_index: Dict[int, List[BinExport2.FlowGraph.Edge]] = defaultdict(list)
+        # edges that end up at the given basic block
+        self.target_edges_by_basic_block_index: Dict[int, List[BinExport2.FlowGraph.Edge]] = defaultdict(list)
+
+        self.vertex_index_by_address: Dict[int, int] = {}
+
+        self.data_reference_index_by_source_instruction_index: Dict[int, List[int]] = defaultdict(list)
+        self.data_reference_index_by_target_address: Dict[int, List[int]] = defaultdict(list)
+        self.string_reference_index_by_source_instruction_index: Dict[int, List[int]] = defaultdict(list)
+
+        self._index_vertex_edges()
+        self._index_flow_graph_nodes()
+        self._index_flow_graph_edges()
+        self._index_call_graph_vertices()
+        self._index_data_references()
+        self._index_string_references()
+
+    def get_basic_block_address(self, basic_block_index: int) -> int:
+        basic_block = self.be2.basic_block[basic_block_index]
+        first_instruction_index = next(self.instruction_indices(basic_block))
+        insn = self.be2.instruction[first_instruction_index]
+        assert insn.HasField("address"), "first insn in a basic block must have an explicit address"
```

Comment:
@williballenthin this assumption does not hold for my test files. See https://github.com/google/binexport/issues/128#issuecomment-2078054664 for additional context.

Reply:
Because of this I'm getting assertion errors. Hopefully we can clear this up with the BinExport maintainers to know for certain how we should handle this.

Reply:
I'm glad I added the assertion then :-)

Yeah, this code is based on the documentation, rather than a large corpus of BinExport2 files.

I concur with your comment on the other thread. I think the conditions should be extended to include "if the instruction is the target of a jump or call, it should have the full address" or something like that.

Reply:
I've fallen back on indexing instruction addresses in [9c99af9](https://github.com/mandiant/capa/pull/1950/commits/9c99af9908773a090f6f2fa52651343cebb3a03d) as a workaround to remove this blocker and continue development and rule writing. The BinExport maintainer needs more time to investigate if BinExport should record the address of the first instruction of all basic blocks. So, for the time being, we can't make this assumption.